chore: responses API sample code (#538)
SAP/ai-sdk: docs-java/foundation-models/openai/responses.mdx

---

> id: responses

Reviewer: (Major)

Please add the new file to the [sidebarsDocsJava.js file](https://github.com/SAP/ai-sdk/blob/8837810e4b37e850b8711af94ad765451fcf3906/sidebarsDocsJava.js#L20) so that is appears in the sidebar of the docs under Foundation Models > OpenAI.

Author: done

title: Responses
hide_title: false
hide_table_of_contents: false

> description: How to use the SAP AI SDK for Java to perform chat completion tasks using OpenAI models deployed on SAP AI Core.

Reviewer 2: ```suggestion
description: How to use the SAP AI SDK for Java to perform responses tasks using OpenAI models deployed on SAP AI Core.
```

:::warning
All classes under any of the `...model` packages are generated from an OpenAPI specification.
This means that these model classes are not guaranteed to be stable and may change with future releases.
They are safe to use, but may require updates even in minor releases.
:::

<details>
<summary id="new-user-interface-v140">New User Interface (v1.4.0)</summary>

We're excited to introduce a new user interface for OpenAI responses calls starting with **version 1.4.0**. This update is designed to improve the SAP AI SDK by:

- **Decoupling Layers:** Separating the convenience layer from the model classes to deliver a more stable and maintainable experience.
- **Staying Current:** Making it easier for the SAP AI SDK to adapt to the latest changes in the OpenAI API specification.
- **Consistent Design:** Aligning with the Orchestrator convenience API for a smoother transition and easier adoption.

**Please Note:**

- The new interface is gradually being rolled out across the SAP AI SDK.
- We welcome your feedback to help us refine this interface.
- The existing interface (v1.0.0) is deprecated but available.

> </details>

Reviewer: (Minor)

I don't think we need this info here anymore.

## Prerequisites

Before using the AI Core module, ensure that you have met all the general requirements outlined in the [overview](../../overview-cloud-sdk-for-ai-java#general-requirements).
Additionally, include the necessary Maven dependency in your project.

### Maven Dependencies

Add the following dependency to your `pom.xml` file:

```xml
<dependencies>
    <dependency>
        <groupId>com.sap.ai.sdk.foundationmodels</groupId>
        <artifactId>openai</artifactId>
        <version>${ai-sdk.version}</version>
    </dependency>

> </dependencies>

Reviewer 2: You are also missing `openai-java-core`, it is [set as optional](https://github.com/SAP/ai-sdk-java/pull/910/changes#diff-b33726c01e2bad5ac82771d33758c198a77631da378a5f2a1c225417cd86abcdR126) in our `openai` module, meaning we [have to add it](https://github.com/SAP/ai-sdk-java/pull/910/changes#diff-21b0333ec3f112e04803b8daa6283e4833d25f3ec163153487ec865793cefc0dR104) to use the client.

  - <details>
    <summary>Example deployed model from the AI Core <code>/deployments</code> endpoint</summary>

    ```json
    {
      "id": "d123456abcdefg",
      "deploymentUrl": "https://api.ai.region.aws.ml.hana.ondemand.com/v2/inference/deployments/d123456abcdefg",
      "configurationId": "12345-123-123-123-123456abcdefg",
      "configurationName": "gpt-4o-mini",
      "scenarioId": "foundation-models",
      "status": "RUNNING",
      "statusMessage": null,
      "targetStatus": "RUNNING",
      "lastOperation": "CREATE",
      "latestRunningConfigurationId": "12345-123-123-123-123456abcdefg",
      "ttl": null,
      "details": {
        "scaling": {
          "backendDetails": {}
        },
        "resources": {
          "backendDetails": {
            "model": {
              "name": "gpt-4o-mini",
              "version": "latest"
            }
          }
        }
      },
      "createdAt": "2024-07-03T12:44:22Z",
      "modifiedAt": "2024-07-16T12:44:19Z",
      "submissionTime": "2024-07-03T12:44:51Z",
      "startTime": "2024-07-03T12:45:56Z",
      "completionTime": null
    }
    ```

    </details>

<!-- prettier-ignore-end-->

> ## Create a non-persistent response

Reviewer: (Major)

Please add a short sentence explaining what the code block does to each of the sections, like you did it for the bottom three.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseCreateParams.builder().input("Hi, how are you?").store(false).build(); // set store to true when creating a persistent response

var response = responseClient.create(params);
```

## Create a persistent response

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseCreateParams.builder().input("Hi, how are you?").store(true).background(true).build(); // set background to true for the response to run asynchronously

var response = responseClient.create(params);
```

> ## Create a streaming non-persistent response

Reviewer: (Question)

Would it make sense to also add an example for streaming+persistence?

Author: Yes it would, I already have a comment which says set store to true if you want to create a persistent response inside the streaming block, I clarified it further.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

> var params = ResponseCreateParams.builder().input("Hi, how are you?").model(ChatModel.GPT_5).store(false).build(); // set store to true when creating a persistent response

Reviewer: (Question)

Is there a reason why you use `.model()` here but not in the other examples? Which one has precedence in the end? I.e., if I use GPT_4 in the `AiCoreOpenAiClient.forModel()` and `.model(ChatModel.GPT_5)` here, which model is used? This should probably be noted to make it clear for users.

Author: Yes, true it's redundant and it is an ignored configuration.

var response = responseClient.createStreaming(params);
```

## Retrieve a response

This allows the retrieval of a previously created persistent response by its id.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseRetrieveParams.builder().responseId("response-id").build();

var response = responseClient.retrieve(params);
```

## Cancel a response

This cancels the previously created background response by its id.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseCancelParams.builder().responseId("response-id").build();

var response = responseClient.cancel(params);
```

## Delete a response

This deletes a stored response by its id.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseDeleteParams.builder().responseId("response-id").build();

> var response = responseClient.delete(params);

Reviewer: `delete()` does not return a value.

```suggestion
responseClient.delete(params);
```

---
id: responses
title: Responses
hide_title: false
hide_table_of_contents: false
description: How to use the SAP AI SDK for Java to perform responses tasks using OpenAI models deployed on SAP AI Core.
keywords:
  - sap
  - cloud
  - sdk
  - ai
---

## Introduction

This guide demonstrates how to use the SAP AI SDK for Java to perform responses using OpenAI models deployed on SAP AI Core.

Refer to the [OpenAI Responses API reference](https://platform.openai.com/docs/api-reference/responses) for details on the underlying OpenAI SDK API.

:::warning
All classes under any of the `...model` packages are generated from an OpenAPI specification.
This means that these model classes are not guaranteed to be stable and may change with future releases.
They are safe to use, but may require updates even in minor releases.

> :::

Reviewer 2: This is not valid right? We don't generate any classes

Author: This is a general warning like the one in chat-completion.mdx regarding model classes.

Reviewer 2: It's not even our package name, I don't see why we should say anything when it clearly belongs to OpenAI, they can document it themselves.

Author: Okay yes understood, but the same exact warning is in chat-completion.mdx, so it was added earlier by someone, do I remove it from both classes now?

## Prerequisites

Before using the AI Core module, ensure that you have met all the general requirements outlined in the [overview](../../overview-cloud-sdk-for-ai-java#general-requirements).
Additionally, include the necessary Maven dependency in your project.

### Maven Dependencies

> Add the following dependency to your `pom.xml` file:

Reviewer 2: ```suggestion
Add the following dependencies to your `pom.xml` file:
```

var response = responseClient.createStreaming(params);
```

## Retrieve a response

This allows the retrieval of a previously created persistent response by its id.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseRetrieveParams.builder().responseId("response-id").build();

var response = responseClient.retrieve(params);
```

## Cancel a response

This cancels the previously created background response by its id.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseCancelParams.builder().responseId("response-id").build();

var response = responseClient.cancel(params);
```

## Delete a response

This deletes a stored response by its id.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params = ResponseDeleteParams.builder().responseId("response-id").build();

responseClient.delete(params);
```

## Useful features using endpoints mentioned above

### Create a background response and poll

> 

Reviewer 2: Why did you add this useful features part?

Author: According to [Jonas's comment here](https://github.com/SAP/ai-sdk/pull/538#pullrequestreview-4542133869) to showcase for users sample code for the features added which used the endpoints.

        .putAdditionalProperty("required", JsonValue.from(List.of("name", "age")))
        .putAdditionalProperty("additionalProperties", JsonValue.from(false))
        .build();

// setting strict to true guarantees the returned JSON conforms exactly to the schema
var format =
    ResponseFormatTextJsonSchemaConfig.builder().name("person").schema(schema).strict(true).build();

var textConfig = ResponseTextConfig.builder().format(format).build();

var params =
    ResponseCreateParams.builder()
        .input("Extract the person: John is 30 years old.")
        .text(textConfig)
        .store(false)
        .build();

var response = responseClient.create(params);
```

### Create a response with truncation

This creates a response with truncation enabled so long conversations are automatically trimmed.

```java
var responseClient = AiCoreOpenAiClient.forModel(GPT_5, "resource-group-name").responses();

var params =
    ResponseCreateParams.builder()
        .input("Can you elaborate on your last point?")
        .previousResponseId("previous-response-id")
        // AUTO drops older items from the middle if the conversation exceeds the context window

>         .truncation(ResponseCreateParams.Truncation.AUTO)

Reviewer 2: truncation is deprecated

Author: but we already merged the code to main, do I remove the sample code from the method here in the documentation?

Reviewer 2: Ok nevermind
